n3fit/src/n3fit/model_trainer.py

```diff
@@ -860,7 +862,9 @@ def hyperparametrizable(self, params):
             # and the seed needs to be updated accordingly
             seeds = self._nn_seeds
             if k > 0:
-                seeds = [np.random.randint(0, pow(2, 31)) for _ in seeds]
+                # seeds = [np.random.randint(0, pow(2, 31)) for _ in seeds]
+                # generate seeds for each k-fold from the input `nnseeds`
+                seeds = [seed * k for seed in seeds]
```

Comment:
Note that the seed can be `0`, so if you'd like the results to be reproducible you may just want to provide a seed for the numpy generator instead.

Reply:
That is true @RoyStegeman; my apologies. I could offer another possibility that seems to work along the lines of generating the same seeds every time.
```
            seeds = self._nn_seeds
            if k > 0:
                seeds_gerator = [np.random.default_rng(seed=seed) for seed in seeds]
                seeds = [generator.random() for generator in seeds_gerator]
```
However, if this condition is not necessary (as you suggested above), I will keep the code as it was previously. Thanks for your very detailed and valuable suggestions

Reply:
I don't think this would work and instead you'd want to do something like `generator.integers(0, pow(2, 31))` as seeds should be integers. Because the seeds for each fold are based on the fixed value `self._nn_seeds` they will still be the same for each fold, is this intentional? Previously you prevented this by multiplying with `k`. 

Either way, feel free to return it as you did originally, I had failed to realize that the `seeds` in this line is already generated using a random number generated and is thus extremely unlikely to be 0 (one way to make it absolutely impossible is by changing the lower bound of that rng to 1 instead of 0).

Reply:
You are completely right @RoyStegeman, my apologies. The correct code would be something like
```
if k > 0:
   seeds_gerator = [np.random.default_rng(seed=seed) for seed in seeds]
   seeds = [generator.integers(1, pow(2, 31)) for generator in seeds_gerator]
```
they do not need to be the same but originated from the same seed.

Reply:
Please have a look at my last commit. Now you'll have a different seed per fold, but same between replicas if the fit is run simultaneously in series. While being different between folds may be helpful in case some seeds end up being particularly good for a given setup and thus doing too well on or hyperopt metric, I don't think they'd need to be different between replicas since the hyperparameters are the things we want to vary in a controlled way. If you'd like the seed to depend on the replica as well, you could probably use `self.replicas` to achieve that. 

Once we start fitting a full distribution we of course do want the different replicas within that distribution to have different seeds, so that is something to keep in mind.

Reply:
Thanks for your help @RoyStegeman. 

I tested here and works perfectly! One thing that was actually raised today during our meeting discussion is whether the seeds (although different in every fold) must also be different for every trial, which is not the case here nor in any of my previous suggestions. So, in summary: should the `seeds = [generator.integers(1, pow(2, 30)) * k for generator in rngs]` also depend (somehow) on the trial number as well ?

In relation to the replica dependence. In the multireplica case I suspect that `self._nn_seeds` (= `nnseeds` argument in `ModelTrainer`) is already a list of seeds for each replica. I am not quite certain but, please, correct if I am wrong. 

Thanks once again for your kind help.

Reply:
> In relation to the replica dependence. In the multireplica case I suspect that self._nn_seeds (= nnseeds argument in ModelTrainer) is already a list of seeds for each replica. I am not quite certain but, please, correct if I am wrong.

This I don't know and personally I'd wait for #1661 before worrying about it, though it's good to keep in mind that we fixed the seed here. 

Between trials of different hyperparameter setups I don't think it's a problem that we re-use the same seed, as long as it's different for fits within that trial. In principle we assume the performance of a fit does not depend on the choice of the seed, in practice there may be certain seeds that are particularly favourable or unfavourable for a given setup but that's something we have to keep in mind even if we have a different seed for each trial.

Reply:
Thanks @RoyStegeman for your clarification. From your thoughts, I suspect that we are OK with the current state of the seeds generation and if so I will proceed with the next part of the PR review by @Radonirinaunimi. Thanks once again for you help and valuable suggestions.